server/base/src/main/java/org/apache/accumulo/server/master/balancer/TabletBalancer.java

```diff
@@ -56,7 +56,11 @@
  * Implementations may wish to store configuration in Accumulo's system configuration using the
  * {@link Property#GENERAL_ARBITRARY_PROP_PREFIX}. They may also benefit from using per-table
  * configuration using {@link Property#TABLE_ARBITRARY_PROP_PREFIX}.
+ *
+ * @deprecated since 2.1.0. Use {@link org.apache.accumulo.core.spi.balancer.TabletBalancer}
+ *             instead.
  */
+@Deprecated(since = "2.1.0")
 public abstract class TabletBalancer {
```

Comment:
Is there any way we can make this implement or extend the SPI version, so we automatically support compatibility with any implementations of the old interface, without two code paths in the manager?

Reply:
I think that could be done. The main downside I can think of is that, when using the deprecated balancer, the params to getAssignments and balance would get converted to the new types in the master and then converted back to the thrift types in the implementation of the SPI TabletBalancer methods that then delegate to the deprecated TabletBalancer methods. I don't think the balancer methods are called often enough that we really need to worry about the performance hit, though.

Reply:
Done in latest commit.

Reply:
> The main downside I can think of is that, when using the deprecated balancer, the params to getAssignments and balance would get converted to the new types in the master and then converted back to the thrift types

@brianloss  I was thinking that AssignmentParamsImpl could be created w/ refs to the new and old types and have methods to get the old types.  The could avoid the conversion from new to old.  I am not sure but it seems like the method that creates AssignmentParamsImpl has refs to the old types.

Reply:
> @brianloss I was thinking that AssignmentParamsImpl could be created w/ refs to the new and old types and have methods to get the old types. The could avoid the conversion from new to old. I am not sure but it seems like the method that creates AssignmentParamsImpl has refs to the old types.

@keith-turner I pushed some changes that I think take care of this.

Reply:
@brianloss  I just took a quick look at, I like how you pulled code out of Master,TabletBalancer and moved it into AssignmentParamsImpl and BalancerParamsImpl.  Making those classes own the responsibility for converting from the SPI types to the internal thrift types is nice.